### ISSUE TITLE
[Bug]: `zfd.checkbox()` should require user to check by default

### ISSUE THREAD

**Issue by TheRealFlyingCoder** [airjp73/remix-validated-form#39]
`zfd.checkbox()` should mean the field is required (e.g accept terms and conditions) to follow zod's pattern

Because your Union transforms undefined into false, technically this is a valid value:

```
export const checkbox = ({ trueValue = "on" }: CheckboxOpts = {}) =>
  z.union([
    z.literal(trueValue).transform(() => true),
    z.literal(undefined).transform(() => false),
  ]);
```

I don't believe the union is necesarry at all, as `undefined` is the same as `false` during checks.

If you skip that transform we can `zfd.checkbox().optional()`

**Comment by airjp73** [comment]
I don't really agree with this. I want `zfd.checkbox()` to be analogous to `z.boolean()`. If the checkbox is checked it should be `true`, and if the checkbox is unchecked then it should be `false`. And like `z.boolean()`, we shouldn't make either value fail the validation by default.

For better or worse, an unchecked checkbox is omitted from the `FormData` when the form is submitted. To me this means there's no meaningful way to use `optional` since there's no way for us to tell the difference between "unchecked" and "there is no checkbox".

That said, you can enforce that a checkbox is checked with `refine`.
```ts
acceptsLicenseAgreement:
  zfd
    .checkbox()
    .refine(val => val, "Please accept the license agreement")
```

**Comment by TheRealFlyingCoder** [comment]
So I think this is a semantical thing.

`z.boolean()` is is a literal, which means TRUE / FALSE are both valid values,

A checkbox on the other hand is a user interaction, where TRUE is the only valid literal, if there is no interaction with the checkbox that doesn't equal FALSE, and that's exactly why it is natively omitted from the FormData.

That's also why radio buttons exist:

o - Accept
o - Decline

This is where TRUE / FALSE are both valid, as it lets us know the user has interacted with the input

**Comment by airjp73** [comment]
The state of a checkbox is also a boolean (checked vs not checked), even if it's not encoded as a boolean type. And it can be checked even if the user hasn't interacted with it (can default to checked).

What I'm ultimately trying to get at with `zfd.checkbox` is "is the checkbox checked, true or false?". It's up to the developer to make the judgement on a per-form basis if the user is required to check that checkbox or not. For the vast majority of forms I've written, both checked an unchecked were totally valid. Every once in awhile you have a form with a checkbox like `acceptsLicenseAgreement` where unchecked is invalid in that particular case.

**Comment by TheRealFlyingCoder** [comment]
Hmm I understand where you are coming from but doesn't it feel weird?

In this scenario the code is going against the expected behaviour by saying:

`zfd.text()` = User must interact and input text (empty string transforms to undefined)
`zfd.text().optional()` = User doesn't have to interact

`zfd.checkbox()` = User doesn't have to interact
`zfd.checkbox().optional()` = Does nothing different
`zfd.checkbox().refine(val => val)` = User must interact

Given that you know undefined = unchecked, i'm not sure how making it false changes anything for the developer other than creating a quirk in the library?

If you need a TRUE/FALSE value you can simply use the key like `let bool = !!data.myCheckbox`

**Comment by airjp73** [comment]
If we don't transform the value to a boolean, then why have a helper at all? It sounds like your ideal version of `checkbox`

* Fails on `undefined` by default
* Doesn't transform to a boolean
* Passes on any string (not just the specified one)

But that's the exact behavior of `z.string`.

```ts
mustBeChecked: z.string(),
canBeUnchecked: z.string().optional(),
coerceToBoolean: z.string().optional().transform(val => val !== undefined)
```

The purpose of having a helper for checkboxes at all, for me, is to get that coercion to happen automatically.

I'm also not super concerned about `optional` not doing anything on `checkbox`. zod surfaces the `optional` method even on schemas where it doesn't make sense to have it (e.g. it's valid to do `z.string().optional().optional()` and the second optional does nothing).

**Comment by TheRealFlyingCoder** [comment]
Sorry for taking a bit to respond! :(

I think the value of a preprocessor turning 'on' => true, and leaving `undefined` as `undefined` by default is a super valuable helper by itself even though it doesn't sound like it's doing much, it's what you'd expect straight out of the gates.

We could easily solve for true / false by letting the dev opt-in when they want `undefined` to equal `false`, something like:

`zfd.checkbox({ trueValue: 'test', useFalse: true })`

You get the best of both worlds that way and no dev's are caught out.

I know it sounds a bit wierd but having `false` in a DB when you don't expect it can **really** matter

**Comment by uhoh-itsmaciek** [comment]
>zfd.checkbox() = User doesn't have to interact
>zfd.checkbox().optional() = Does nothing different
>zfd.checkbox().refine(val => val) = User must interact

This is only true if you assume the checkbox is unchecked by default. Many are checked by default (e.g., "Please subscribe me to your newsletter" on signup). I would find an unchecked value failing validation by default to be surprising.

**Comment by TheRealFlyingCoder** [comment]
> This is only true if you assume the checkbox is unchecked by default. Many are checked by default (e.g., "Please subscribe me to your newsletter" on signup)

A native `<input type="checkbox" />` is unchecked by default so that's what you'd assume... just because the developer can set defaults doesn't change that? 

> I would find an unchecked value failing validation by default to be surprising.

Like i've said above, semantically checkbox can only have 1 value, this helper should just be turning that value from `string` => `true` for simplicity ootb.

Adding the second value of `false` while useful in some scenarios is going against the native behaviour of FormData, hence why I suggested it being an 'opt-in' behaviour.

**Comment by airjp73** [comment]
I'm going to close this because I still think the current behavior is desirable. If you think we could clarify any gotchas in the [checkbox api reference](https://www.remix-validated-form.io/zod-form-data/api-reference#checkbox), I'd be happy to look at a PR for that.

**Comment by uhoh-itsmaciek** [comment]
@TheRealFlyingCoder thinking about this again and how checkboxes work in forms specifically, I kinda see your point. I now think either behavior makes sense, depending on how you map your thinking about forms to the conceptual "schema" defined for user inputs. I do think your 'useFalse' mechanism is kind of unclear, but there's probably a different way to make that work.

Sorry if I my feedback contributed to this issue getting closed. For what it's worth, I withdraw my objection to changing this @airjp73 (though again, I think either behavior is reasonable).

**Comment by airjp73** [comment]
Sorry, I never really properly responded to @TheRealFlyingCoder 's last comment. 😄 

> Adding the second value of false while useful in some scenarios is going against the native behaviour of FormData, hence why I suggested it being an 'opt-in' behaviour.

I'm not really sure how concerned this library should be with preserving the native behavior of FormData. The whole point is to smooth over the native behavior and allow you to think more about the data differently.

A good example is `zfd.text`. You don't really need it
```ts
// with zfd.text
required: zfd.text(),
optional: zfd.text(z.string().optional()),

// without zfd.text
required: z.string().nonempty(),
optional: z.string(),
```

But I'd rather think in terms of the empty text input having no value (undefined) vs being an empty string. Which is why `zfd.text` transforms empty strings to `undefined`. This is not the native behavior of FormData.

> You get the best of both worlds that way and no dev's are caught out.

I think returning `true | undefined` is going to catch more people out than returning a `boolean`. The data _is_ a boolean -- there are only two possible values no matter which way we do it. The option with the least cognitive overhead is going to be modeling the data as an actual `boolean`.

> I know it sounds a bit wierd but having false in a DB when you don't expect it can really matter, just as undefined is a necessary use case a lot of the time.

You could make the same DB argument in reverse about having `null` in a DB when you want `false` causing problems. The type definitions and documentation are clear (hopefully) about what's going to be returned.

---

> Sorry if I my feedback contributed to this issue getting closed. For what it's worth, I withdraw my objection to changing this @airjp73 (though again, I think either behavior is reasonable).

No need to worry @uhoh-itsmaciek! Additional feedback is always welcome. 😄

**Comment by TheRealFlyingCoder** [comment]
>Sorry, I never really properly responded to @TheRealFlyingCoder 's last comment. 😄

All good! :P I also kind of wish we could just magically get some more points of view on this, I know I might not be right, but I think it's worth a serious discussion haha

>I'm not really sure how concerned this library should be with preserving the native behavior of FormData. 

My opinion for helper libraries it's more about touching the least you need to by default... Preserving the native behaviour is just a bonus.

>I think returning true | undefined is going to catch more people out than returning a boolean. 

So I still don't see how this is possible given the context of zod.

By default `zfd.checkbox()` would fail parsing saying "required"... Devs that know a checkbox is undefined by default would expect that, and if you didn't know, you understand how zod works, so you should realise that it's undefined and add `.optional()` for a checkbox that doesn't need to be checked.

>You could make the same DB argument in reverse about having null in a DB when you want false causing problems

This would be true except that the dev has to opt in to allowing `undefined` with `.optional()`, otherwise it would fail validation. 

This is my key point, where you shouldn't create a value from nothing unless the dev has opted in, and clearly understands that it's happening

**Comment by airjp73** [comment]
As a side-note. Before publishing this library this was actually called `booleanCheckbox` and a renamed it for conciseness. But it seems like maybe that was the better choice. 😆

**Comment by kiliman** [comment]
Kind of jumping in here by request of @TheRealFlyingCoder. I see where @airjp73 is coming from. However, it's not uncommon to need a "tri-value" switch. Basically `true|false|undefined`.

In my app, we have questions that could be yes or no, but we want to know if it's explicitly no, not that they forgot to check it. Therefore I actually render that as a multi-segment button [Yes][No][Unanswered] where unanswered is the default.

So basically what I'm saying is, use a checkbox where you want to get true if it's checked, and only allow `false|undefined` if `optional`.

If you need to support `true|false|optional`, then don't use checkbox.

**Comment by airjp73** [comment]
Maybe this is a use-case thing. Most of the checkboxes I've written are configuration. "User has permission to do X", "enable X feature", "allow nonessential cookies", etc. None of this information is _optional_ it's just true or false. Even when the checkbox is unchecked it still communicates information.

I think I see where others are coming from. Checkboxes are often used in that "please check this box" kind of way and `optional` means they don't _have to_ check the box -- they're allowed to leave it unchecked.

So maybe this is also a preference/perspective thing. If there's a strong desire for a helper that functions in the way that @TheRealFlyingCoder describes, I'm open to discussing ways that can fit in the library as a separate helper. And if we need to we can rename this helper too.

**Comment by TheRealFlyingCoder** [comment]
>In my app, we have questions that could be yes or no, but we want to know if it's explicitly no, not that they forgot to check it.

@kiliman this is a common scenario for me as well, especially for legality in forms, and is the main distinction here:

Checkbox = true / undefined <-- "Agree" is the only action
Radio button = true / false / undefined  <--  User can "Agree" or "Disagree" (Or any number of options)

`false` inherently means "The user said no".
`undefined` inherently means "The user didn't choose anything".

That's a super important distinction for form validation ^ 

@airjp73 It's easy enough to cover both scenarios, it should just be opt-in not opt-out is all... Start with just turning 'on' -> `true` and let the dev decide if they want to transform undefined to false depending on their scenario.

You could have a seperate helper for that type of checkbox scenario, or just add it as a flag

**Comment by airjp73** [comment]
> false inherently means "The user said no".
undefined inherently means "The user didn't choose anything".

With a checkbox, there's no way to meaningfully tell the difference between these two scenarios. I'd rather have the helper return whether or not the checkbox is checked and have the developer derive meaning / add validation on top of that.

**Comment by TheRealFlyingCoder** [comment]
We are definitely going around in circles haha

Is there a reason you don't want to just do both? That solves the whole discussion :P 

Just adding a flag for turning `undefined` => `false` will let you switch it on whenever you speciifcally need `false` as a value instead of `undefined`?

**Comment by airjp73** [comment]
Doing both would mean two separate helpers, imo. Adding a configuration option clutters the api and adds almost as much noise as doing `zfd.checkbox().optional().transform(Boolean)`. And whether it's a configuration option or a second helper, I feel like the library loses some coherency somehow. If I do end up "doing both", I feel like I would need a really good story for when & why a developer should choose one over the other.

I think the real sticking point for me is that the current behavior works for all use-cases already. And the differences between that and the proposed behavior feel more and more like personal preference the more we talk about it.

If you feel like the api as-is has some gotchas, I'm open to suggestions on how to improve the docs to help people avoid those.

**Comment by TheRealFlyingCoder** [comment]
>I think the real sticking point for me is that the current behavior works for all use-cases already. And the differences between that and the proposed behavior feel more and more like personal preference the more we talk about it.

The proposal is based off the expected behaviour of FormData + Zod, if you were to do it without the helper? I raised it after being immediately caught out, and wondering why it wasn't validating.

If there are valid reasons why `false` makes more sense, than I wouldn't question it, but so far we haven't discussed any that `undefined` doesn't also work for

**Comment by TheRealFlyingCoder** [comment]
Just saw your edit, breaking changes are good to get in early :P

**Comment by kiliman** [comment]
Like I said, if I need to model `true|false|undefined` as valid values, then I would **not** chose a checkbox. It can only reasonably return `truthy|falsy` values based on check (it either is or isn't).

So my suggestion is to not change checkbox. `zfd.checkbox()` would mean check required like *Accept agreement*. `zfd.checkbox().optional()` would be for settings like *Remember password*.

Instead add another helper, whether it be a radio group or toggle buttons, etc. Something that allows a choice between multiple values.

**Comment by airjp73** [comment]
@kiliman, your `remix-params-helper` lib behaves the same way this does currently, right? The transform for `z.boolean()` coerces the value of the checkbox into a boolean but doesn't add any validation on top? That's essentially my goal here. The helper says "this is a boolean" and the dev can decide if it needs to be true or false.

**Comment by kiliman** [comment]
True, but in my case, I'm simply defining the types of the results. I don't have a checkbox helper to actually render a checkbox.

The input helpers are just there to add some props. It's up to the user to create the inputs for whatever UI they want. The default for `boolean` is `type=checkbox`, but you could override the type as well.

```ts
const ActionSchema = z.object({
  accept: z.boolean(),
})

<label>
  <div>Accept (must be checked):</div>
  <input {...inputProps('accept')} />
</label>
```

Because I didn't include `optional()`, it's adding the `required` attribute to the input.

![image](https://user-images.githubusercontent.com/47168/151570482-9d8db790-1655-4afa-ab38-f45c19f23b2a.png)

**Comment by kiliman** [comment]
```ts
const ActionSchema = z.object({
  accept: z.boolean(),
  remember: z.boolean().optional(),
  answer: z.boolean().optional(),  
})

<label>
  <div>Accept (must be checked):</div>
  <input {...inputProps('accept')} />
</label>
<label>
  <div>Remember Password:</div>
  <input {...inputProps('remember')} />
</label>
<label>
  <div>Question:</div>
  <div style={{ display: 'flex', gap: '16px' }}>
  {/* override checkbox type to radio */}
  <label><input {...inputProps('answer')} type="radio" value="true" /> Yes</label>
  <label><input {...inputProps('answer')} type="radio" value="false" /> No</label>
  <label><input {...inputProps('answer')} type="radio" value="" /> Not answered</label>
  </div>
</label>
```
![image](https://user-images.githubusercontent.com/47168/151576787-86a9fdae-a090-444c-9c8d-d9264ac757e2.png)

The helper will convert the `true|false|undefined` values for `answer` automatically.

See the sample app at https://remix-params-helper.herokuapp.com/demos/actions

**Comment by airjp73** [comment]
Ah, I see. Thanks for clearing that up! :)

My personal preference would be to construct schema a little more explicitly, but I see I don't think your approach is wrong.

```ts
const ActionSchema = zfd.formData({
  accept: zfd.checkbox(),
  remember: zfd.checkbox().refine(val => val),
  answer: z.union([
    z.literal("yes"),
    z.literal("no"),
    z.literal("unanswered")
  ]),
})
```

I can definitely see that `.refine(val => val)` is a little less ergonomic than `.optional()` in the alternate api.

## Path forward

I would accept a PR that adds a helper like this
```ts
// Probably need to handle `trueValue` too, but I think `errorMessage` has a higher priority and more common use case
const pleaseCheck = (errorMessage?: string) =>
  z
    .literal("on", { required_error: errorMessage })
    .transform(() => true); // transform is optional -- you don't even need the value in your action since you know it's always true
```

This suggestion feels like it:
* Has a clear story and distinct use-case for when it should be used over `checkbox`
  * The name `pleaseCheck` is pretty clear but could go for alternate names that communicate the same thing
* Has an ergonomic api for its use-case

For me, I want a helper that does what `zfd.checkbox` does. I'm open to discussing alternate names in a separate issue and I'll take a PR for an alternate helper as well. If you want to hash out the details of the alternate helper before making a PR, we could discuss that in a separate issue as well.

**Comment by kiliman** [comment]
Yeah, I think your library and my helper are like two different sides of the same coin.

I'm more focused on validating the results from FormData, URLSearchParams, etc. against my schema (View Model). It's up to the UI to send the data in the correct format. I added helpers to help with some of the boilerplate props.

Whereas your library renders the UI elements for you based on your form schema. So your stuff is probably a little more opinionated than mine, but at the same time you get more out of the box.

Hope that makes sense.

**Comment by TheRealFlyingCoder** [comment]
>So my suggestion is to not change checkbox. zfd.checkbox() would mean check required like Accept agreement. zfd.checkbox().optional() would be for settings like Remember password.

@kiliman  I'm a little confused with this sentence... because that is the change we are discussing, or do you mean that checkbox shouldn't be changing `undefined`? :P 

Right now:
`zfd.checkbox()` is optional
`zfd.checkbox().refine(val => val)` is required

>I would accept a PR that adds a helper like this

@airjp73 If you were going to add a helper it would be to add `undefined` => `false` because that's the extra bit of work being done?

It doesn't make sense to do extra work first, then add a helper to roll it back.

**Comment by airjp73** [comment]
@kiliman, Yeah, they're definitely two approaches to the same problem.

I want to disambiguate a little since this is now a monorepo with a few different, related packages. The `checkbox` helper lives in the `zod-form-data` package which is a pretty small, focused package.

`zod-form-data` directly tackles the same problem (server-side) as `remix-params-helper`: validating FormData and URLSearchParams against a zod schema. But it doesn't do anything with the UI side of things. It can be used separately from `remix-validated-form` but works in harmony with it as well.

The difference in the approach is that, instead of having a helper that crawls the schema and decides how to transform the raw data, the user builds their schema with helpers that have a built-in `z.preprocess` to transform the data. This makes it work with automatic nested data and array support built into `remix-validated-form`, and also ensures that we can support more complex zod schemas like where the whole schema is one big union.

`remix-validated-form` is validation-library-agnostic (can use yup or something else), and handles client & server validation in a similar manor to Formik or react-hook-form. It doesn't really render the ui for you so much as it allows you to build components that integrate automatically with your schema.

---

@TheRealFlyingCoder 

> The developer can add any extra stuff they want on top

This also applies to the current version of `checkbox`

> Other than being a breaking change, whats the aversion to making it the new checkbox?

`checkbox` works as intended and I don't think the behavior incorrect, just different from your desired approach.

**Comment by TheRealFlyingCoder** [comment]
I understand that the `checkbox` works as you intended, and that there are ways to get back to native behaviour but that's normally the wrong way to approach a library... especially a remix one where our whole credence is "use the web as it's intended" :P 

Right now I have 3 active projects that i've brought zfd into, and none of them **require** a `false` value over `underfined`, but all of them have multiple required checkboxes.

So for myself alone, i've had to `zfd.checkbox().refine(val => val)` several times already and it's been a week haha

Have you had any scenario's where you actually needed the `false` value? This is the biggest question... Because I can't remember a time when i've needed it

**Comment by airjp73** [comment]
> Have you had any scenario's where you actually needed the false value? This is the biggest question... Because I can't remember a time when i've needed it

We're going in circles again haha. I'm honestly not sure I've ever written a checkbox where the user was required to check it. If I have it was only a couple. For context, I've worked on mostly heavy enterprise apps with a lot of configuration pages. Public-facing websites are bound to have more of these "please check" types of checkboxes.

**Comment by TheRealFlyingCoder** [comment]
We are definitely going around in circles 😄, but this is my main point:

if you don't have a scenario where `false` is necessary, you've essentially solved for something that was never a problem, and created confusion at the same time.

All checkbox needed to do was turn a string into true behind the scenes, and then we get a nice schema that makes sense:

```
const ActionSchema = zfd.formData({
  name: zfd.text(),
  description: zfd.text().optional
  terms: zfd.checkbox(),
  remember: zfd.checkbox().optional()
})
```

**Comment by airjp73** [comment]
In the interest of not spiraling on this forever I think this is going to have to be my last comment on this issue. I've outlined in [this comment](https://github.com/airjp73/remix-validated-form/issues/39#issuecomment-1024404087) what I see as the path forward for this.

> if you don't have a scenario where false is necessary, you've essentially solved for something that was never a problem, and created confusion at the same time.

I think this library should take form data and return properly typed and well modeled representation of the data in the form. From a data modeling perspective, `boolean` is a more accurate representation of a checkbox than `true | undefined`. `true | undefined` is also a "boolean", it's just not a `boolean` which makes it confusing.

**Comment by TheRealFlyingCoder** [comment]
Yeah honestly i'm very tired of this thread haha it's clear to me that the heels have dug in
